tezedge_state/src/effects/effects_recorder.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeSet, VecDeque};
+use std::fmt::{self, Debug};
+
+use crypto::nonce::Nonce;
+
+use crate::peer_address::{PeerAddress, PeerListenerAddress};
+use crate::Effects;
+
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Default, Clone)]
+pub struct RecordedEffects {
+    nonces: VecDeque<Nonce>,
+    chosen_peers_to_connect_to: VecDeque<Vec<PeerListenerAddress>>,
+    chosen_potential_peers_for_advertise: VecDeque<Vec<PeerListenerAddress>>,
+    chosen_potential_peers_for_nack: VecDeque<Vec<PeerListenerAddress>>,
+}
+
+impl RecordedEffects {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl Effects for RecordedEffects {
+    fn get_nonce(&mut self, _: &PeerAddress) -> Nonce {
+        self.nonces
+            .pop_front()
+            .expect("no more recorded nonces avaiable")
```

Comment:
@binier 
we cannot use expect in state machine, this is unrecoverable

Reply:
That is while replaying recording. We should panic if state machine requested nonce more times at current run than while recording. That means non-determinism and we should panic to detect it.

Reply:
@binier 
well, I am just afraid if anybody can still use it accidentally for something other for running node,
I know it will mess a nice trait Effects api without Result :),
at least put there, please, some comments at the begining
```
//! Never use, dedicated just for replay...
```

maybe, it should be aslo some feature/cfg (never used it before)